### PR TITLE
helm3: init at 3.0.0 (help wanted)

### PR DESCRIPTION
**Pull request by novoxd**
###### Motivation for this change
Add helm3 version. This version is made as a separate package because helm of version 3 is completely different from version 2 (no tiller, different templates and logic, different go dependencies ). From the user view, there is a need for both versions at the same time, cause most of Kubernetes clusters work with helm2, and new ones with helm3 at the same time. The `default.nix` is copied from the original helm package with modifications.

**Help wanted with rename output binary to helm3.**  I am new to nixpkgs (its my 2nd package, 1st isn't even merged yet). As said upper - both packages helm in nixpkgs (verions2) and helm3 needed. But with this `default.nix` this package produces executable output as 'helm' what conflicts with old helm package. I tried to rename output in different hooks like `fixupPhase`, and overriding `installPhase`  but with no success. I think its quite simple to do, but I can't google anything or find in mans. 
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

P.S. This is my first package build and first pr to nixpkgs, sorry if something wrong.

### CODE REVIEW

**Comment by rlupton20** [comment]
From the looks of the `buildGoPackage` implementation

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-packages/generic/default.nix

you might want to try implementing a preInstall hook. I'm not terribly familiar with the infrastructure around building Go packages, so I might be a ways off!

File: pkgs/applications/networking/cluster/helm3/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    # This is a hack(?) to flatten the dependency tree the same way glide or dep would
+    # Otherwise you'll get errors like
+    # have DeepCopyObject() "k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime".Object
+    # want DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
+    rm -rf $NIX_BUILD_TOP/go/src/k8s.io/kubernetes/vendor
+    rm -rf $NIX_BUILD_TOP/go/src/k8s.io/apiextensions-apiserver/vendor
+  ''; 
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/kubernetes/helm;
```
Comment (doronbehar) [inline pkgs/applications/networking/cluster/helm3/default.nix:38]: https://github.com/NixOS/rfcs/pull/45

Please quote the url.

File: pkgs/applications/networking/cluster/helm3/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  pname = "helm3";
+
+  src = fetchFromGitHub {
+    owner = "helm";
+    repo = "helm";
+    rev = "v${version}";
+    sha256 = "0gx5gmj1470q3gj8v043dmm31skf83p1ckzdcfzx8wdjlglsljrj";
+  };
+
+  modSha256 = "06599w4rvipb669vc5rj62k2kjakz2vmpp270carf204qbm5grk8";
```
Comment (doronbehar) [inline pkgs/applications/networking/cluster/helm3/default.nix:14]: Rebase on top of latest master and switch to `vendorSha256`.

File: pkgs/applications/networking/cluster/helm3/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  goPackagePath = "helm.sh/helm/v3";
```
Comment (doronbehar) [inline pkgs/applications/networking/cluster/helm3/default.nix:16]: I _think_ this is not needed.

File: pkgs/applications/networking/cluster/helm3/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  subPackages = [ "cmd/helm"];
+
+  #goDeps = ./deps.nix;
```
Comment (doronbehar) [inline pkgs/applications/networking/cluster/helm3/default.nix:19]: Remove.

**Review by doronbehar — CHANGES_REQUESTED** [review]

**Comment by doronbehar** [comment]
> Help wanted with rename output binary to helm3. I am new to nixpkgs (its my 2nd package, 1st isn't even merged yet). As said upper - both packages helm in nixpkgs (verions2) and helm3 needed. But with this default.nix this package produces executable output as 'helm' what conflicts with old helm package. I tried to rename output in different hooks like fixupPhase, and overriding installPhase but with no success. I think its quite simple to do, but I can't google anything or find in mans.

Yea it's a bit hard to explain why this failed, and it's definitely not ideal that you've had this experience - it shouldn't be that hard and not intuitive + undocumented. However, I once saw this working for someone:

https://github.com/NixOS/nixpkgs/pull/86927/files#diff-2663ae5d05323a4fc069b6f0f3e664fbR20

Question: If:
> From the user view, there is a need for both versions at the same time, cause most of Kubernetes clusters work with helm2, and new ones with helm3 at the same time.

Then what's so wrong about having it colliding with the helm2 ?

**Comment by teto** [comment]
dont we have already kubernetes-helm in the repo ? should we close this ?

**Comment by yurrriq** [comment]
> should we close this ?

yep

**Pull request closed without merging**